### PR TITLE
test(features): add bdd for existing ci-tests

### PR DESCRIPTION
**Pull request by sinhaashish**
## Pull Request template

Please, go through these steps before you submit a PR.

**Why is this PR required? What issue does it fix?**:

**What this PR does?**:

**Does this PR require any upgrade changes?**:

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Any additional information for your reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/website is used to track them:

**PLEASE REMOVE BELOW INFORMATION BEFORE SUBMITTING**

The PR title message must follow convention:
   `<type>(<scope>): <subject>`.

Where: <br />
- `type` is defining if release will be triggering after merging submitted changes, details in [CONTRIBUTING.md](../CONTRIBUTING.md).
    Most common types are:
    - `feat`      - for new features, not a new feature for build script
    - `fix`       - for bug fixes or improvements, not a fix for build script
    - `chore`     - changes not related to production code
    - `docs`      - changes related to documentation
    - `style`     - formatting, missing semi colons, linting fix etc; no significant production code changes
    - `test`      - adding missing tests, refactoring tests; no production code change
    - `refactor`  - refactoring production code, eg. renaming a variable or function name, there should not be any significant production code changes

- `scope` is a single word that best describes where the changes fit.
  Most common scopes are like:
    - data engine (`localpv`, `jiva`, `cstor`)
    - feature (`provisioning`, `backup`, `restore`, `exporter`)
    - code component (`api`, `webhook`, `cast`, `upgrade`)
    - test (`tests`, `bdd`)
    - chores (`version`, `build`, `log`, `travis`)

- `subject` is a single line brief description of the changes made in the pull request.

### CODE REVIEW

File: tests/bdd/zfs.feature [hunk 1]
```diff
@@ -0,0 +1,87 @@
+Feature: Validate volume provisioning for fsType zfs, ext4, xfs, btrfs
+         Validate snapshot creation and clone
+         Validate volume resize
+
+  Background:
+    Given a single zfs cluster is configured
+    And zfs pool pool created on the node
+
+  Scenario: volume provision for file system type zfs, ext4, xfs, btrfs
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs respectively
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: Use scenario outline here?

File: tests/bdd/zfs.feature [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs respectively 
+    And pvc is created referencing the same sc
+    When a deployment is created using the same pvc
+    Then zfsvolume must be created and used by the application
+
+  Scenario: volume property change for file system type zfs, ext4, xfs, btrfs
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs respectively 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the zfsvolume properties like compression, dedep are updated
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: dedep --> dedup

File: tests/bdd/zfs.feature [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    Then the zfsvolume properties must reflect the updated value
+
+  Scenario: volume property change for file system type zfs, ext4, xfs
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: btrfs missing from here? Better to use scenario outline for these everywhere

File: tests/bdd/zfs.feature [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs respectively 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the zfsvolume volume type is DATASET and record size is updated
+    Then the record size of must show the updated value of zfsvolume
+
+  Scenario: volume property change for file system type zfs, ext4, xfs
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: We cannot use same scenario name for all the cases. Please change it to meaningful ones.

File: tests/bdd/zfs.feature [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    And a zfsvolume is created and used by the application
+    When the snapshot create command for the given pvc is run
+    Then the snapshot must be created for the corresponding pvc
+
+  Scenario: resize a pvc when fsType is zfs, ext4, xfs
+    Given a storage class is created with  zfs, ext4, xfs separately 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the size of pvc is updated 
+    Then the pvc must relfect the new size
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: Add a step to validate the underlying dataset size as well.

File: tests/bdd/zfs.feature [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    When the zfsvolume volume type is not DATASET and block size is updated
+    Then the block size of zfsvolume must be updated
+
+  Scenario: create a snaphot from a pvc
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs separately 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    And a zfsvolume is created and used by the application
+    When the snapshot create command for the given pvc is run
+    Then the snapshot must be created for the corresponding pvc
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: Step needed for verifying underlying dataset.

File: tests/bdd/zfs.feature [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+  Scenario: resize a pvc when fsType is zfs, ext4, xfs
+    Given a storage class is created with  zfs, ext4, xfs separately 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the size of pvc is updated 
+    Then the pvc must relfect the new size
+
+  Scenario: create a clone from a snaphot
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: snaphot --> snapshot

Comment (Abhinandan-Purkait) [reply tests/bdd/zfs.feature]: Please correct below as well.

File: tests/bdd/zfs.feature [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    Then the record size of must show the updated value of zfsvolume
+
+  Scenario: volume property change for file system type zfs, ext4, xfs
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs respectively 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the zfsvolume volume type is not DATASET and block size is updated
+    Then the block size of zfsvolume must be updated
+
+  Scenario: create a snaphot from a pvc
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: snaphot --> snapshot

File: tests/bdd/zfs.feature [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    Given a storage class is created with  zfs, ext4, xfs separately 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the size of pvc is updated 
+    Then the pvc must relfect the new size
+
+  Scenario: create a clone from a snaphot
+    Given a snaphot for a pvc
+    When the clone create command is run for the given snaphot 
+    Then the clone must be created from the snapshot
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: Step needed to validate the underlying dataset size as well.

File: tests/bdd/zfs.feature [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+  Scenario: create a clone  from a volume
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: Extra space here.

File: tests/bdd/zfs.feature [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    Given a pvc
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: Specify the properties as well. We need to test snapshot and clone with all volume properties

File: tests/bdd/zfs.feature [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    When the clone create command is run for the given volume 
+    Then the clone must be created from the volume
+
+  Scenario: volume provision for a raw block volume
+    Given a storage class is created without any fstype as Raw block volume does not have any fstype
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: raw block specification is not a part of sc params. It comes from pvc. Please update it accordingly.

File: tests/bdd/zfs.feature [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    And pvc is created referencing the same sc
+    When a deployment is created using the same pvc
+    Then zfsvolume must be created and used by the application
+
+  Scenario: volume property for a raw block volume
+    Given a storage class is created without any fstype as Raw block volume does not have any fstype
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the zfsvolume properties like compression, dedep are updated
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: dedep --> dedup

File: tests/bdd/zfs.feature (hunk 12 above)
Comment (Abhinandan-Purkait) [reply tests/bdd/zfs.feature]: Update other cases as well

File: tests/bdd/zfs.feature [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    Then the zfsvolume properties must reflect to the updated value
+
+  Scenario: volume property for a raw block volume
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: Again same scenario name cannot be used. Please update with the expected scenario.

File: tests/bdd/zfs.feature [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    And a deployment is created using the same pvc
+    When the zfsvolume volume type is DATASET and record size is updated
+    Then the record size of zfsvolume must be updated
+
+  Scenario: volume property for a raw block volume
+    Given a storage class is created without any fstype as Raw block volume does not have any fstype
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the zfsvolume volume type is not DATASET and block size is updated
+    Then the block size of zfsvolume must be updated
```
Comment (Abhinandan-Purkait) [inline tests/bdd/zfs.feature]: These can be just scenario outlines. Why do we need multiple cases?

**Review by Abhinandan-Purkait — COMMENTED** [review]
I think we need to combine the scenarios that have same set of actions but just different set of properties are used, they should be a scenario outline.

File: tests/bdd/zfs.feature [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+
+  Background:
+    Given a single zfs cluster is configured
+    And zfs pool pool created on the node
+
+  Scenario: volume provision for file system type zfs, ext4, xfs, btrfs
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs respectively 
+    And pvc is created referencing the same sc
+    When a deployment is created using the same pvc
+    Then zfsvolume must be created and used by the application
```
Comment (w3aman) [inline tests/bdd/zfs.feature]: zfsvolume creation should not be dependent on any deployment creation using pvc. So we can try to write it in this way. WDYT ?

Given a storage class is created with fsType as <<(fsType)>>
And a pvc is created referencing this storage class
Then pvc should be in bound state 
And zfsvolume must be created and should be in Ready state to use by any application
When a application deployment is created using this pvc
Then deployment pod should be in Running state

File: tests/bdd/zfs.feature [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the zfsvolume volume type is DATASET and record size is updated
+    Then the record size of must show the updated value of zfsvolume
+
+  Scenario: volume property change for file system type zfs, ext4, xfs
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs respectively 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    When the zfsvolume volume type is not DATASET and block size is updated
```
Comment (w3aman) [inline tests/bdd/zfs.feature]: instead of volume type is not DATASET, we can write volume type is ZVOL

File: tests/bdd/zfs.feature [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+    Then the block size of zfsvolume must be updated
+
+  Scenario: create a snaphot from a pvc
+    Given a storage class is created with fsType as zfs, ext4, xfs, btrfs separately 
+    And pvc is created referencing the same sc
+    And a deployment is created using the same pvc
+    And a zfsvolume is created and used by the application
+    When the snapshot create command for the given pvc is run
```
Comment (w3aman) [inline tests/bdd/zfs.feature]: may be good idea to add that snapshot should be in same namespace as pvc

Comment (w3aman) [reply tests/bdd/zfs.feature]: zfssnapshot resource should also be check for ready status.state

File: tests/bdd/zfs.feature [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+    Then the clone must be created from the snapshot
+
+  Scenario: create a clone  from a volume
+    Given a pvc
+    When the clone create command is run for the given volume 
+    Then the clone must be created from the volume
+
+  Scenario: volume provision for a raw block volume
+    Given a storage class is created without any fstype as Raw block volume does not have any fstype
+    And pvc is created referencing the same sc
```
Comment (w3aman) [inline tests/bdd/zfs.feature]: using volumeMode as Block

File: tests/bdd/zfs.feature (hunk 11 above)
Comment (sinhaashish) [reply tests/bdd/zfs.feature]: the test just creates a deployment which used the clone. No properties are verified.

**Review by w3aman — APPROVED** [review]
LGTM

**Review by Abhinandan-Purkait — APPROVED** [review]

**Pull request merged**